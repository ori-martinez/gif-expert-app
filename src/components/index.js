/* Exportaci√≥n Conjunta de los Componentes */
export * from './AddCategory';
export * from './GifCard';
export * from './GifsGrid';
