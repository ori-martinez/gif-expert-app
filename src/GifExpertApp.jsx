import { useState } from 'react';
/* Componentes */
import { AddCategory, GifsGrid } from './components';

// ESTILOS
/* Textos */
const h1Style = "p-2 m-4 text-center text-4xl text-teal-600 font-bold";

// TAREAS
/* __________________________________________________________________________________________________________
    
    1. Hacer la función addCategory para agregar una categoria al clickear un botón
    2. Usar el setCategories para añadir un nuevo elemento en categories desde el componente handleCategory
    3. Mostrar en pantallas el título de los gifs de la categoría buscada, usando el id de estos como key, en
el componente GifsGrid
    4. Hacer pruebas sobre el componente GifExpertApp
__________________________________________________________________________________________________________ */

// COMPONENTE
/* Búscador de Gifs por Categorías */
export const GifExpertApp = () => {
    // CONSTANTES
    /* Estado de las Categorías Buscadas */
    const [ categories, setCategories ] = useState([ 'Fullmetal Alchemist' ]);

    // FUNCIONES
    /* Manejador de Inserción de una Categoría */
    const handleCategory = (newCategorie) => {
        const upperCategories = categories.map(category => category.toUpperCase());

        if (upperCategories.includes(newCategorie.toUpperCase())) return;

        setCategories([ newCategorie, ...categories ]);
    }

    // RETORNO
    return (
        <>
            { /* Título */ }
            <h1 className={ h1Style }> GifExpert App </h1>

            { /* Input de Búsqueda */ }
            <AddCategory handleCategory={ handleCategory } />

            { /* Listado de Gifs */ }
            {categories.map(category => 
                <GifsGrid 
                    key={ category }
                    gifsCategory={ category }
                    gifsQty={ 9 }
                />
            )}
        </>
    );
}

/* __________________________________________________________________________________________________________

NOTAS:
    1. Los Functional Components nunca deben tener la ejecución directa de una función, ya que estas se
volverán a ejecutar cada vez que se rendericen dichos componentes 
__________________________________________________________________________________________________________ */
