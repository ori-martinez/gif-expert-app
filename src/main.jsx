import React from 'react';
import ReactDOM from 'react-dom/client';
/* Estlos */
import './index.css';
/* Componentes */
import { GifExpertApp } from './GifExpertApp';

ReactDOM.createRoot(document.getElementById('root')).render(
    <React.StrictMode>
        <GifExpertApp />
    </React.StrictMode>
);
