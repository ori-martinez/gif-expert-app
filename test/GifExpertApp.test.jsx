import { fireEvent, render, screen } from '@testing-library/react';
/* Components */
import { GifExpertApp } from '../src/GifExpertApp';

// CONSTANTES
/* Valor Deseado del Elemento input */
const inputValue = 'Harry Potter';

describe('Pruebas del <GifExpertApp />', () => {
    // PRUEBA DE COINCIDENCIA
    test('Debe hacer match con el snapshot', () => {
        /* Renderización del Componente */
        const { container } = render(<GifExpertApp />);

        expect(container).toMatchSnapshot();
    });

    // PRUEBA DE COINCIDENCIA DEL ESTADO INICIAL DE categories
    test('La categoría inicial debe ser Fullmetal Alchemist', () => {
        /* Renderización del Componente */
        render(<GifExpertApp />)

        expect(screen.getByText('Fullmetal Alchemist'));
    });

    // PRUEBA DE FUNCIONALIDAD CON UNA CATEGORIA NUEVA
    test('La función handleCategory() debe agregar una nueva categoría', () => {
        /* Renderización del Componente */
        render(<GifExpertApp />)

        // CONSTANTES
        /* Elementos input y form */
        const input = screen.getByRole('textbox');
        const form = screen.getByRole('form');

        /* Simulación del Evento Input */
        fireEvent.input(input, { target: { value: inputValue } });
        /* Simulación del Evento Submit */
        fireEvent.submit(form);

        expect(screen.getByText(inputValue));
    });

    // PRUEBA DE FUNCIONALIDAD CON UNA CATEGORIA REPETIDA
    test('La función handleCategory() no debe agregar una categoría repetida', () => {
        /* Renderización del Componente */
        render(<GifExpertApp />)

        // CONSTANTES
        /* Elementos input y form */
        const input = screen.getByRole('textbox');
        const form = screen.getByRole('form');

        // CATEGORIA NUEVA
        /* Simulación del Evento Input */
        fireEvent.input(input, { target: { value: inputValue } });
        /* Simulación del Evento Submit */
        fireEvent.submit(form);
        // CATEGORIA REPETIDA
        /* Simulación del Evento Input */
        fireEvent.input(input, { target: { value: inputValue } });
        /* Simulación del Evento Submit */
        fireEvent.submit(form);

        expect(screen.getAllByText(inputValue).length).not.toBe(2);
    });
})
