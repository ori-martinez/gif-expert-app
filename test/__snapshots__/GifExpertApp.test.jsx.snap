// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del <GifExpertApp /> Debe hacer match con el snapshot 1`] = `
<div>
  <h1
    class="p-2 m-4 text-center text-4xl text-teal-600 font-bold"
  >
     GifExpert App 
  </h1>
  <form
    aria-label="form"
    class="flex justify-center"
  >
    <input
      class="p-1 pl-2 my-4 w-2/5 text-lg text-stone-900 border border-stone-900 rounded-lg placeholder:text-stone-400 focus:outline-none focus:ring focus:ring-violet-300"
      placeholder="Search gifs..."
      type="text"
      value=""
    />
  </form>
  <div
    class="mx-auto mt-6 w-4/5"
  >
    <h3
      class="ml-12 p-3 text-3xl text-teal-500 font-bold"
    >
       
      Fullmetal Alchemist
       
    </h3>
  </div>
  <h2
    class="text-center text-4xl text-stone-900 font-bold"
  >
     Cargando... 
  </h2>
  <div
    class="mx-auto my-4 w-4/5 grid grid-cols-3 gap-12 rounded-lg"
  />
</div>
`;
